RISC V vector predication support intrinsics support

zvookin opened the pull request:
Have RISC V use LLVM vector predication intrinsics and add support for some architecture specific intrinsics.

steven-johnson commented on src/CodeGen_RISCV.cpp:
```diff
@@ -97,7 +187,188 @@ int CodeGen_RISCV::target_vscale() const {
+    builder->SetInsertPoint(block);
+
+    // Set vector fixed-point rounding flag if needed for intrinsic.
+    bool round_down = intrin.flags & RISCVIntrinsic::RoundDown;
```
Do we need explicit `() != 0` here? The current flag values are fine, but if we ever get bits above the first 8, not sure if the language promises this to work :-/

zvookin replied:
Conversion to ```bool``` is a comparison against zero. If the value is zero, it's false, otherwise true. Normally I would add the comparison explicitly, but for this usage, it feels idiomatic to treat the ```&``` operator as set inclusion. As in I feel adding the explicit test makes the code harder to read, not easier.

rootjalex commented on src/CodeGen_RISCV.cpp:
```diff
@@ -1,24 +1,114 @@
+    enum {
+        AddVLArg = 1 << 0,          // Add a constant full size vector length argument
+        RoundDown = 1 << 1,         // Set rounding mode to down (rdn) before intrinsic.
+        RoundUp = 1 << 2,           // Set rounding mode to up (rdu) before intrinsic.
```
Could you provide a link in the comments that points to documentation of the rounding modes?

zvookin replied:
Not in inline comments as it's pretty long and to a PDF so I can't (easily) link directly to the section anyway. Strikes me as the sort of thing where it's pretty obvious what they do and one has to know a bit about the RISC V vector architecture to work on the code at all. I added "vxrm" to make it clear these are the fixed-point rounding mode.

steven-johnson commented on src/CodeGen_RISCV.cpp:
```diff
@@ -97,7 +187,188 @@ int CodeGen_RISCV::target_vscale() const {
+        // TODO: When LLVM finally fixes the instructions to take rounding modes,
+        // this will have to change to passing the rounding mode to the intrinsic.
+        llvm::FunctionType *csrw_llvm_type = llvm::FunctionType::get(void_t, {xlen_type}, false);
+        llvm::InlineAsm *inline_csrw = llvm::InlineAsm::get(csrw_llvm_type, "csrw vxrm,${0:z}", "rJ,~{memory}", true);
```
...I had no idea you could do inline asm this way in LLVM!

zvookin replied:
It introduces a fairly significant code motion barrier, at least for this usage, and thus is highly undesirable. I"m pretty sure this is what inline assembly in C/C++ turns into in LLVM IR.

steven-johnson commented on src/CodeGen_RISCV.cpp:
```diff
@@ -97,7 +187,188 @@ int CodeGen_RISCV::target_vscale() const {
+    }
+
+    // Call the LLVM intrinsic.
+    // TODO: This does not work for reductions.
```
Should we open an Issue for this? Or add `internal_assert(not-a-reduction);`?

zvookin replied:
Reductions are now handled through the vector predication intrinsics, hence the reason I had that TODO there is obsolete. (It was tracking that I needed to figure out a way to make it work.) There are likely more to be handled, but it doesn't really need an assert as there's no way to specify an intrinsic that does a reduction in this part of the code. This is a thing to consider when revamping ```call_intrin```, but I don't think it needs a tracking issue as there is an ```is_reduction``` argument now. (I.e. I probably fixed the underlying issue there too.)

steven-johnson commented on src/CodeGen_RISCV.cpp:
```diff
@@ -97,7 +187,188 @@ int CodeGen_RISCV::target_vscale() const {
+    wrapper->addFnAttr(llvm::Attribute::ReadNone);
+    wrapper->addFnAttr(llvm::Attribute::NoUnwind);
+
+    // llvm::verifyFunction(*wrapper);
```
Is this meant to be commented out?

steven-johnson reviewed (commented):
Looks good, mostly nits

rootjalex commented on src/CodeGen_RISCV.cpp:
```diff
@@ -97,7 +187,188 @@ int CodeGen_RISCV::target_vscale() const {
+    {"vwsub", {Type::Int, 2}, "widening_sub", {Type::Int, Type::Int}, RISCVIntrinsic::AddVLArg | RISCVIntrinsic::MangleReturnType},
+    {"vwsubu", {Type::UInt, 2}, "widening_sub", {Type::UInt, Type::UInt}, RISCVIntrinsic::AddVLArg | RISCVIntrinsic::MangleReturnType},
+    {"vwmul", {Type::Int, 2}, "widening_mul", {Type::Int, Type::Int}, RISCVIntrinsic::AddVLArg | RISCVIntrinsic::MangleReturnType},
+    {"vwmulu", {Type::UInt, 2}, "widening_mul", {Type::UInt, Type::UInt}, RISCVIntrinsic::AddVLArg | RISCVIntrinsic::MangleReturnType},
```
Does LLVM currently offer an intrinsic for `vwmulsu`?

zvookin replied:
Added. Wasn't sure if there is normalization of the argument ordering and didn't see it happening anywhere so I added support for handling either order.

rootjalex replied:
There isn’t on main. I found it useful to add normalization on a project branch I’m working on, I’ll probably make a PR to add that and clean up some patterns later.

zvookin commented on src/CodeGen_RISCV.cpp:
```diff
@@ -97,7 +187,188 @@ int CodeGen_RISCV::target_vscale() const {
+
+    // Produce intrinsic name and
+    std::vector<llvm::Type *> llvm_arg_types;
+    std::string mangled_name = "llvm.riscv.";
```
Because that will not compile without casting one of the ```const char *``` values to ```std::string``` and if I have to mention the type again anyway, this strikes me as a bit clearer.

steven-johnson replied:
Gotcha

zvookin commented on src/CodeGen_RISCV.cpp:
```diff
@@ -1,24 +1,114 @@
 
+constexpr int max_intrinsic_args = 4;
+
+struct FixedOrScalableVector {
```
I changes the name to something that is not much shorter but is more descriptive.

rootjalex commented:
@steven-johnson Do the buildbots test RISC-V cross-compilation?

If so, this PR should probably make an addition to simd_op_check, to check for the added vector instructions in this PR.

steven-johnson commented:
> @steven-johnson Do the buildbots test RISC-V cross-compilation?
> 
> If so, this PR should probably make an addition to simd_op_check, to check for the added vector instructions in this PR.

Not currently, no. We should definitely add that.

abadams commented on src/CodeGen_RISCV.cpp:
```diff
@@ -97,7 +187,188 @@ int CodeGen_RISCV::target_vscale() const {
+        internal_assert(!(round_down && round_up));
+        llvm::Value *rounding_mode = llvm::ConstantInt::get(xlen_type, round_down ? 2 : 0);
+        // TODO: When LLVM finally fixes the instructions to take rounding modes,
+        // this will have to change to passing the rounding mode to the intrinsic.
```
Is there an existing LLVM issue for it? Having to use inline assembly to use any of the ops that rely on a rounding flag seems ... unfinished.

zvookin commented:
Added an issue for simd_op_check coverage. https://github.com/halide/Halide/issues/7122

zvookin commented:
Yeah, I'd rant about not having the rounding mode flags in the intrinsic at
the start is badly broken, but it's sort of how things are with LLVM and
RISC V. Really they ought to be in the opcode, but you know, bits are
expensive. I put in a reference to the one PR on LLVm and will dig up
another in a bit.

On Tue, Oct 25, 2022 at 2:09 PM Andrew Adams ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/CodeGen_RISCV.cpp
> <https://github.com/halide/Halide/pull/7119#discussion_r1004966935>:
>
> > +    llvm::Function *wrapper =
> +        llvm::Function::Create(wrapper_ty, llvm::GlobalValue::InternalLinkage,
> +                               wrapper_name, module.get());
> +    llvm::BasicBlock *block =
> +        llvm::BasicBlock::Create(module->getContext(), "entry", wrapper);
> +    llvm::IRBuilderBase::InsertPoint here = builder->saveIP();
> +    builder->SetInsertPoint(block);
> +
> +    // Set vector fixed-point rounding flag if needed for intrinsic.
> +    bool round_down = intrin.flags & RISCVIntrinsic::RoundDown;
> +    bool round_up = intrin.flags & RISCVIntrinsic::RoundUp;
> +    if (round_down || round_up) {
> +        internal_assert(!(round_down && round_up));
> +        llvm::Value *rounding_mode = llvm::ConstantInt::get(xlen_type, round_down ? 2 : 0);
> +        // TODO: When LLVM finally fixes the instructions to take rounding modes,
> +        // this will have to change to passing the rounding mode to the intrinsic.
>
> Is there an existing LLVM issue for it? Having to use inline assembly to
> use any of the ops that rely on a rounding flag seems ... unfinished.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/halide/Halide/pull/7119#discussion_r1004966935>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALOUFEADYCC7HI3PIWVT6LWFBEARANCNFSM6AAAAAARNTLZIE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

zvookin commented:
Had to do a number of bug fixes on the vector predication work as I messed up on the test coverage previously. It works better with the code in this branch...

zvookin commented:
I believe I have addressed the feedback raised. In a couple cases, by opening tracking issues, but mostly by incorporating the changes suggested or fixing the problems raised. If I've missed anything please let me know. I also did some better testing. Obviously that should get rolled into the real tests, which is effectively the simd_op_check issue. This should be ready to land.

steven-johnson reviewed (approved):
LGTM pending green

rootjalex commented on src/CodeGen_RISCV.cpp:
```diff
@@ -97,7 +187,188 @@ int CodeGen_RISCV::target_vscale() const {
+    if (effective_vscale != 0) {
+        for (const RISCVIntrinsic &intrin : intrinsic_defs) {
+            if (intrin.ret_type.type_pattern == FixedOrScalableVector::WildcardWidths) {
+                for (int scale = 0; scale < 4; scale++) {
```
@zvookin a comment stating something like “iterate through 8, 16, 32, 64 bit return types” could be helpful here

rootjalex reviewed (approved):
Also LGTM with a nit

zvookin commented on src/CodeGen_RISCV.cpp:
```diff
@@ -97,7 +187,188 @@ int CodeGen_RISCV::target_vscale() const {
+                llvm::Function *intrin_impl = define_riscv_intrinsic_wrapper(intrin, 1);
+                Type ret_type = concretize_fixed_or_scalable(intrin.ret_type, 1,
+                                                             target.vector_bits);
+                std::vector<Type> arg_types;
```
Refactored to avoid allocating the vector for each intrinsic.
